src/query/pipeline/core/src/finished_chain.rs

```diff
@@ -0,0 +1,388 @@
+    }
+
+    pub fn apply(&mut self, mut info: ExecutionInfo) -> Result<()> {
+        let chain = std::mem::take(&mut self.chain);
+
+        let mut states = Vec::with_capacity(chain.len());
+        let mut callbacks = Vec::with_capacity(chain.len());
+        let mut always_callbacks = Vec::with_capacity(chain.len());
+
+        for (location, callback) in chain.into_iter() {
+            if !callback.always_call() {
+                callbacks.push((location, callback));
+            } else {
+                always_callbacks.push((location, callback));
+            }
+        }
+
+        for (location, callback) in callbacks {
+            let instant = Instant::now();
+            if let Err(cause) = callback.apply(&info) {
+                states.push(ApplyState {
+                    location,
+                    is_always: false,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                });
+
+                info.res = Err(cause.clone());
+                Self::apply_always(info, states, always_callbacks);
+
+                return Err(cause);
+            }
+
+            states.push(ApplyState {
+                location,
+                is_always: false,
+                successfully: true,
+                elapsed: instant.elapsed(),
+            });
+        }
+
+        Self::apply_always(info, states, always_callbacks);
+        Ok(())
+    }
+
+    fn apply_always(
+        info: ExecutionInfo,
+        mut states: Vec<ApplyState>,
+        always_callbacks: Vec<(&'static Location<'static>, Box<dyn Callback>)>,
+    ) {
+        for (location, always_callback) in always_callbacks {
+            let instant = Instant::now();
+            states.push(match always_callback.apply(&info) {
+                Ok(_) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: true,
+                    elapsed: instant.elapsed(),
+                },
+                Err(_cause) => ApplyState {
+                    location,
+                    is_always: true,
+                    successfully: false,
+                    elapsed: instant.elapsed(),
+                },
+            });
+        }
+
+        Self::log_states(&states);
+    }
+
+    pub fn extend(&mut self, other: FinishedCallbackChain) {
+        self.chain.extend(other.chain)
+    }
+
+    fn log_states(apply_states: &[ApplyState]) {
+        let mut message = String::new();
+        writeln!(&mut message, "Executor apply finished callback state:").unwrap();
+        for apply_state in apply_states {
+            let execute_state = match apply_state.successfully {
+                true => "\u{2705}",
+                false => "\u{274C}",
+            };
+
+            let always_state = match apply_state.is_always {
+                true => "(always) ",
+                false => "",
+            };
+
+            writeln!(
```

Comment:
just a suggestion, I prefer to require a name for each `callback`, just like the processors, make it more readable.

and then we can display it by adding  just one line when display a pipeline ?

Reply:
Great suggestion. but it is only recorded in the log. Is the caller location okay?

Reply:
may ok if there are no more than 3 or 4 callbacks.
some can be identified from the file path if the debugger has seen it before.